src/invoke/mod.rs

```diff
@@ -12,36 +12,41 @@ use super::{
 // Invoking behaves very differently depending on what is invoked
 pub fn invoke<'a>(
   data_stack: &mut Vec<DataValue>,
-  program_stack: &mut Vec<ProgramValue>,
+  program_queue: &mut VecDeque<ProgramValue>,
   local_scope: &mut HashMap<String, RealValue>,
   global_scope: &mut HashMap<String, RealValue>,
   built_in_functions: &HashMap<&'a str, &'a dyn BuiltInFunction>,
 ) {
-  match match data_stack.pop() {
+  let value = match data_stack.pop() {
     Some(DataValue::Real(v)) => v,
     Some(DataValue::Label(l)) => {
       if let Some(v) = local_scope.get(&l) { v.clone() }
       else if let Some(v) = global_scope.get(&l) { v.clone() }
+      else if built_in_functions.contains_key(&l.as_str()) { RealValue::BuiltInFunction(l) }
       else { panic!("Undefined label dereference: {}", l); }
     },
     None => panic!("Invoked on empty data_stack!"),
-  } {
-    // Invoking a substack puts it on your program_stack and resumes execution
+  };
+  match value {
+    // Invoking a substack puts it on your program_queue and resumes execution
     // (This means it executes in current context / has access to local scope)
-    RealValue::Substack(mut s) => {
-      program_stack.append(&mut s);
+    RealValue::Substack(s) => {
```

Comment:
There has to be a better way to do this, without making a dequeue out of `s`? Or at least that doesn't shuffle the data back and forth between `program_queue` and `s_queue`, there should be an append_to_front operation or similar?

Reply:
(In my previous adventures trying to make large changes to vectors I have used `split_off` a lot to avoid the time complexity of repeated inserts.
https://doc.rust-lang.org/std/collections/struct.VecDeque.html#method.split_off
https://github.com/sidju/add-ed/blob/main/src/old_buffer/editing.rs#L63

Not sure how applicable it is to this, but thought it related enough to warrant mention.)

Reply:
I had thought so too, especially since it's a ring_buffer under the hood. Might not have looked close enough.

But was kinda banking on that the rust compiler would solve it (The answer to anything that sounds too annoying) Like it should just be to "change the pointer" from s_queue to program_queue. Perhaps an "=" would convey that better?